homeassistant/components/amcrest/__init__.py

```diff
@@ -184,6 +200,38 @@ def _wrap_test_online(self, now):
             pass
 
 
+def _monitor_events(hass, name, api, event_codes):
+    event_codes = ",".join(event_codes)
+    start_stop = re.compile(r"Code=([^;]+);action=(Start|Stop)", flags=re.S)
+    while True:
+        api.available_flag.wait()
+        try:
+            for event_info in api.event_stream(event_codes, retries=5):
+                _LOGGER.debug("%s event info: %r", name, event_info)
+                for code, action in start_stop.findall(event_info):
+                    signal = service_signal(SERVICE_EVENT, name, code)
+                    start = action == "Start"
+                    _LOGGER.debug("Sending signal: '%s': %s", signal, start)
+                    dispatcher_send(hass, signal, start)
+        except AmcrestError as error:
+            _LOGGER.warning(
+                "Error while processing events from %s camera: %r", name, error
+            )
+
+
+def _start_event_monitor(hass, name, api, event_codes):
+    global _event_threads  # pylint: disable=invalid-name
```

Comment:
```suggestion
    global _event_threads  # pylint: disable=invalid-name
```

@bdraco is right here, don't use global.

Reply:
Can someone explain why?

Reply:
It is more of a generally programming concept. I'm pretty darn sure if you search for "Why are globals bad/evil" you will find tons of good reasons.

Don't get me wrong, we have some globals in the core of Home Assistant, they have (rare) use cases; Other than that, we aren't using it anywhere.

Reply:
Thanks, but I've been a programmer for a long time. I know the trade-offs of globals. Just because they can be bad in some or even most uses cases doesn't make them bad in all. But if globals shouldn't be used, then isn't `hass.data` effectively a global, too? And a very public one at that? Much more public than a module level global with a leading underscore. I fail to see why that would be better. What am I missing?

As a compromise, would it be better if I converted those functions into a Class (possibly a subclass of `threading.Thread`), turning the global into a class variable?

Reply:
> Thanks, but I've been a programmer for a long time. I know the trade-offs of globals.

Sorry m8, it was not a personal attack in any way. I just meant to answer a question that consisted of 4 words. I cannot read or extract any level of knowledge or other possible emotions from that. So forgive me for providing a general answer, I mean no harm.

`hass` isn't global, although I understand the common problem you are raising in this use case. The difference here would be that `hass.data[DOMAIN]` is specially meant for storing run-time application data (in a sense globally, but in a single place, namespaced, and passed in and documented). IMO is better than the growth of globally defined variables everywhere throughout the codebase.

However, that can become a long opinionated discussion 😉 

> As a compromise, would it be better if I converted those functions into a Class (possibly a subclass of threading.Thread), turning the global into a class variable?

Yeah, that might work, handling the monitoring of events in a separate class.

Reply:
We should use hass.data over global. That way we can guarantee that things are global on a per-hass instance basis.

However, in this case, why not use the `name` arg as part of the name ? That would be the most recognizable I think.

Reply:
Some background about the `hass.data` instead of `global` decision: our testing infrastructure would end up relying on all these accidental left around globals, would break if the test order would change and leak memory. It was a lot of pain but we got rid of most globals. 

You will still find some globals but those should be cleaned up one day. It's the same with asking people to put things in PyPI packages. We first need to stop it from happening, then we can start reducing it to eventually eliminate it. 

We have blocked pull requests because integrations first needed to have existing code be put into a PyPI package before we were able to accept new pull requests for it. 

The goal of the PyPI rule is for reusability and maintainability. PyPI packages can be reused by other projects which will result in improved quality and less bugs, flourishing the open source ecosystem. The 2nd reason is that reviewers are really overloaded and we should not have to go through PRs that contain protocol specific information but instead focus on the integration of it in Home Assistant.

In this case, I think that the `event_stream` method should emit event objects/dictionaries that can be consumed by Home Assistant. This will improve the re-usability of the library. 

I see that the [`amcrest` package](https://github.com/tchellomello/python-amcrest) was started by Marcelo. I know him personally and also recently contributed to his Ring package. Pretty sure that there will be a quick turnaround time.

Reply:
@balloob, thanks for the explanation. That makes a lot of sense.

Yes, I could use the `name` arg in constructing the thread's name, which would obviate the need for a counter. Since these names are not constrained in any way (other than being unique from other amcrest camera names), can I use them as-is (e.g., "`Amcrest Kitchen IPC`"), or should I massage them in some way? I.e., would you want the thread names to meet some sort of naming convention? FWIW, this processing would only need to happen once during startup for each camera.

Regarding the amcrest package, I was given write permission quite some time ago and have done most of the recent work, especially for use by Home Assistant. I could certainly push more of the processing down into the package where it belongs. But as I've said, if I do it for this, then I should really be doing it for all the other commands/methods used by HA. And I would have to do it in such a way as to maintain backwards compatibility for all the other use cases of that package. E.g., I'd probably have to add a layer on top of most of the existing commands, but that just seems like a band-aid. And there are several other major changes that should probably be made (such as not embedding communications in property methods, etc.) I've considered forking the project and doing a major overhaul to bring it more in line with HA expectations, dropping Python2 support, cli, etc., but that is a significant project. Although I agree it's not optimal, is it ok to leave this as-is for now, especially since it is consistent with the way the rest of the package is used, and then maybe tackle the overhaul later?

Reply:
We have no rules for thread names, no massaging necessary.

I think that it's not necessary to fork the project and revamp all CLI commands just to integrate this. I would just do a breaking change bump to 2.0 and just include this one thing.